compiler/rustc_lint/src/map_unit_fn.rs

```diff
@@ -0,0 +1,114 @@
+use crate::{LateContext, LateLintPass, LintContext};
+
+use rustc_hir::{Closure, Expr, ExprKind, QPath, Stmt, StmtKind};
+use rustc_middle::query::Key;
+use rustc_middle::ty;
+
+declare_lint! {
+    /// The `mapping_to_unit` lint checks for `Iterator::map` receive
+    /// a callable that returns `()`.
+    ///
+    /// ### Example
+    ///
+    /// ```rust
+    /// fn foo(items: &mut Vec<u8>) {
+    ///     items.sort();
+    /// }
+    ///
+    /// fn main() {
+    ///     let mut x: Vec<Vec<u8>> = vec![
+    ///         vec![0, 2, 1],
+    ///         vec![5, 4, 3],
+    ///     ];
+    ///     x.iter_mut().map(foo);
+    /// }
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// Mapping to `()` is almost always a mistake.
+    pub MAP_UNIT_FN,
+    Warn,
+    "`Iterator::map` call that discard the iterator's values"
+}
+
+declare_lint_pass!(MapUnitFn => [MAP_UNIT_FN]);
+
+impl<'tcx> LateLintPass<'tcx> for MapUnitFn {
+    fn check_stmt(&mut self, cx: &LateContext<'tcx>, stmt: &Stmt<'_>) {
+        if stmt.span.from_expansion() {
+            return;
+        }
+
+        if let StmtKind::Semi(expr) = stmt.kind {
+            if let ExprKind::MethodCall(path, receiver, args, span) = expr.kind {
+                if path.ident.name.as_str() == "map" {
+                    if receiver.span.from_expansion()
+                        || args.iter().any(|e| e.span.from_expansion())
+                        || !is_impl_slice(cx, receiver)
+                    {
+                        return;
+                    }
+                    let ty = cx.typeck_results().expr_ty(&args[0]);
+                    let mut fn_name = "";
+                    if let ExprKind::Path(qpath) = &args[0].kind {
+                        if let QPath::Resolved(_, path) = qpath {
+                            fn_name = path.segments[0].ident.name.as_str();
+                        }
+                    }
+                    if let ty::FnDef(id, _) = ty.kind() {
+                        let fn_ty = cx.tcx.fn_sig(id).skip_binder();
+                        let ret_ty = fn_ty.output().skip_binder();
+                        if ret_ty.is_unit() || ret_ty.is_never() {
+                            cx.emit_spanned_lint(
+                                MAP_UNIT_FN,
+                                span,
+                                MappingToUnit {
+                                    function_label: cx.tcx.span_of_impl(*id).unwrap(),
+                                    argument_label: args[0].span,
+                                    map_label: ty.default_span(cx.tcx),
+                                    suggestion: span,
+                                    replace: format!("for_each({})", fn_name),
+                                },
+                            )
+                        }
+                    } else if let ExprKind::Closure(Closure { def_id, fn_decl, body, .. }) =
+                        args[0].kind
```

Comment:
You handle 2 cases: when we have a function name, and when we have an inline closure.
What about the case of a closure name?
```
let f = |_| {};
iter.map(f);
```
Should we prefer having a fully type-based reasoning here?